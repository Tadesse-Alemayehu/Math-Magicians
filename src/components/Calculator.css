#Calculator {
  --orange: #f5913e;
  --light: #e9e8e8;
  --dark: #858693;

  max-width: max-content;
  display: grid;
  grid-template-columns: auto auto;
  grid-template-areas: 'result result';
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  font-weight: 400;
}

#result {
  grid-area: result;
  color: white;
  font-size: 1.2rem;
  background-color: transparent;
  background-color: var(--dark);
}

#result input {
  text-align: right;
  height: 100%;
  width: 95%;
  outline: none;
  display: block;
  margin: auto;
  border: 0;
  padding: 5px;
}

#leftSide {
  display: grid;
  grid-template-columns: auto auto auto;
  grid-template-areas: 'zero zero';
  background-color: var(--light);
}

#rightSide {
  display: grid;
  grid-template-columns: auto;
  background-color: var(--orange);
}

button {
  padding: 1vw 3vw;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: transparent;
}

#rightSide button {
  padding: 1vw 5vw;
}

.zero {
  grid-area: zero;
  grid-row: 5;
}

#Calculator * {
  border: 0.1px rgb(255, 255, 255) solid;
}

.operator {
  background-size: 15px;
  background-repeat: no-repeat;
  background-position: center;
}