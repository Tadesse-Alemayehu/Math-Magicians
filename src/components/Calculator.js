import React, { PureComponent } from 'react';
import './Calculator.css';
import Button from './button';
import calculate from '../logic/calculate';

export default class Calculator extends PureComponent {
  buttonName = null;

  constructor(props) {
    super(props);
    this.state = {
      total: 0,
      next: 0,
      operation: '',
    };
  }

  componentDidMount() {
    console.log('component is mounted');
    this.buttonName = 'AC';
  }

  /**
 * Given a button name and a calculator data object, return an updated
 * calculator data object.
 *
 * Calculator data object contains:
 *   total:s      the running total
 *   next:String       the next number to be operated on with the total
 *   operation:String  +, -, etc.
 */
  calculateNewData = (event) => {
    console.log(event.target.textContent);
    const newData = calculate(
      this.state,
      event.target.textContent,
    );
    console.log(newData);
    // console.log(this.state);
    // this.setState(newData);
    // console.log(this.state);
    // call the calculator method with this.state and button
  }

  updateNextNumber = (event) => {
    this.setState((state) => ({
      total: state.total,
      next: state.next + event.target.textContent,
      operation: state.operation,
    }));

    console.log('new state is', this.state);
  }

  updateOperator = (event) => {
    // console.log(event.target);
    this.setState((state) => ({
      total: state.total,
      next: state.next,
      operation: event.target.textContent,
    }));
    console.log('new state is', this.state);
  }

  render() {
    const { total } = this.state;
    return (
      <div id="Calculator">
        <div id="result">
          <input type="text" name="result" id="result" value={total} />
        </div>
        <div id="leftSide">
          <Button value="AC" buttonName="AC" clickCallback={this.calculateNewData} />
          <Button value="+/-" buttonName="+/-" clickCallback={this.calculateNewData} />
          <Button value="%" buttonName="%" clickCallback={this.updateOperator} />
          <Button value="1" clickCallback={this.updateNextNumber} />
          <Button value="2" clickCallback={this.updateNextNumber} />
          <Button value="3" clickCallback={this.updateNextNumber} />
          <Button value="4" clickCallback={this.updateNextNumber} />
          <Button value="5" clickCallback={this.updateNextNumber} />
          <Button value="6" clickCallback={this.updateNextNumber} />
          <Button value="7" clickCallback={this.updateNextNumber} />
          <Button value="8" clickCallback={this.updateNextNumber} />
          <Button value="9" clickCallback={this.updateNextNumber} />
          <Button value="0" className="zero" clickCallback={this.updateNextNumber} />
          <Button value="." clickCallback={this.updateNextNumber} />
        </div>
        <div id="rightSide">
          <Button className="divide operator" value="รท" clickCallback={this.updateOperator} />
          <Button className="multiply operator" value="x" clickCallback={this.updateOperator} />
          <Button className="minus operator" value="-" clickCallback={this.updateOperator} />
          <Button className="plus operator" value="+" clickCallback={this.updateOperator} />
          <Button className="equal operator" value="=" clickCallback={this.calculateNewData} />
        </div>
      </div>
    );
  }
}
