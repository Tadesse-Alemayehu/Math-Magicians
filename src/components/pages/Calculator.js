import React, { useState, useEffect } from 'react';
import './Calculator.css';
import Button from '../../logic/button';
import calculate from '../../logic/calculate';

export default function Calculator() {
  const [total, setTotal] = useState(null);
  const [next, setNext] = useState(null);
  const [operation, setOperation] = useState(null);
  const [result, setResult] = useState(null);

  useEffect(() => {
    setTotal(0);
    setNext(0);
    setOperation('');
    setResult(0);
  }, []);

  const calculateNewData = (event) => {
    console.log({ total, next, operation }, event.target.textContent);
    const newData = calculate({ total, next, operation }, event.target.textContent);
    setTotal(newData.total || 0);
    setNext(newData.next || 0);
    setOperation(newData.operation || '');
    setResult(newData.total || newData.next || 0);
    console.log(newData);
  };

  const updateNextNumber = (event) => {
    const newResult = result === 0
      ? event.target.textContent : result + event.target.textContent;
    const nextNumber = next === 0
      ? event.target.textContent : next + event.target.textContent;
    setNext(nextNumber);
    setResult(newResult);
  };

  const updateOperator = (event) => {
    const newData = calculate({ total, next, operation }, event.target.textContent);
    setTotal(newData.total || 0);
    setNext(0);
    setOperation(event.target.textContent);
    setResult((newData.total || '0') + event.target.textContent);
  };
  useEffect(() => {
  });
  return (
    <div id="container">
      <div id="textContent">
        <h1 className="mainHeading">Math Magicians</h1>
        <h2>Let&apos;s do some math!</h2>
      </div>
      <div id="Calculator">
        <div id="result">
          <input type="text" name="result" id="result" value={result} />
        </div>
        <div id="leftSide">
          <Button value="AC" buttonName="AC" clickCallback={calculateNewData} />
          <Button value="+/-" buttonName="+/-" clickCallback={calculateNewData} />
          <Button value="%" buttonName="%" clickCallback={updateOperator} />
          <Button value="7" clickCallback={updateNextNumber} />
          <Button value="8" clickCallback={updateNextNumber} />
          <Button value="9" clickCallback={updateNextNumber} />
          <Button value="4" clickCallback={updateNextNumber} />
          <Button value="5" clickCallback={updateNextNumber} />
          <Button value="6" clickCallback={updateNextNumber} />
          <Button value="1" clickCallback={updateNextNumber} />
          <Button value="2" clickCallback={updateNextNumber} />
          <Button value="3" clickCallback={updateNextNumber} />
          <Button value="0" className="zero" clickCallback={updateNextNumber} />
          <Button value="." clickCallback={updateNextNumber} />
        </div>
        <div id="rightSide">
          <Button className="divide operator" value="รท" clickCallback={updateOperator} />
          <Button className="multiply operator" value="x" clickCallback={updateOperator} />
          <Button className="minus operator" value="-" clickCallback={updateOperator} />
          <Button className="plus operator" value="+" clickCallback={updateOperator} />
          <Button className="equal operator" value="=" clickCallback={calculateNewData} />
        </div>
      </div>
    </div>
  );
}
