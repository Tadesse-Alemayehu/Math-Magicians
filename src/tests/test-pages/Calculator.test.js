import renderer from 'react-test-renderer';
import Calculator from '../../components/pages/Calculator';

it('Calculator matches snapshot', () => {
  const tree = renderer
    .create(<Calculator />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
