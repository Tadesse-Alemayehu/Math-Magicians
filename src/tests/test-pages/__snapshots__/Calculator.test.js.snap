// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator matches snapshot 1`] = `
<div
  id="container"
>
  <div
    id="textContent"
  >
    <h1
      className="mainHeading"
    >
      Math Magicians
    </h1>
    <h2>
      Let's do some math!
    </h2>
  </div>
  <div
    id="Calculator"
  >
    <div
      id="result"
    >
      <input
        id="result"
        name="result"
        type="text"
        value={null}
      />
    </div>
    <div
      id="leftSide"
    >
      <button
        name="AC"
        onClick={[Function]}
        type="button"
      >
        AC
      </button>
      <button
        name="+/-"
        onClick={[Function]}
        type="button"
      >
        +/-
      </button>
      <button
        name="%"
        onClick={[Function]}
        type="button"
      >
        %
      </button>
      <button
        name="7"
        onClick={[Function]}
        type="button"
      >
        7
      </button>
      <button
        name="8"
        onClick={[Function]}
        type="button"
      >
        8
      </button>
      <button
        name="9"
        onClick={[Function]}
        type="button"
      >
        9
      </button>
      <button
        name="4"
        onClick={[Function]}
        type="button"
      >
        4
      </button>
      <button
        name="5"
        onClick={[Function]}
        type="button"
      >
        5
      </button>
      <button
        name="6"
        onClick={[Function]}
        type="button"
      >
        6
      </button>
      <button
        name="1"
        onClick={[Function]}
        type="button"
      >
        1
      </button>
      <button
        name="2"
        onClick={[Function]}
        type="button"
      >
        2
      </button>
      <button
        name="3"
        onClick={[Function]}
        type="button"
      >
        3
      </button>
      <button
        className="zero"
        name="0"
        onClick={[Function]}
        type="button"
      >
        0
      </button>
      <button
        name="."
        onClick={[Function]}
        type="button"
      >
        .
      </button>
    </div>
    <div
      id="rightSide"
    >
      <button
        className="divide operator"
        name="รท"
        onClick={[Function]}
        type="button"
      >
        รท
      </button>
      <button
        className="multiply operator"
        name="x"
        onClick={[Function]}
        type="button"
      >
        x
      </button>
      <button
        className="minus operator"
        name="-"
        onClick={[Function]}
        type="button"
      >
        -
      </button>
      <button
        className="plus operator"
        name="+"
        onClick={[Function]}
        type="button"
      >
        +
      </button>
      <button
        className="equal operator"
        name="="
        onClick={[Function]}
        type="button"
      >
        =
      </button>
    </div>
  </div>
</div>
`;

exports[`renders correctly 1`] = `
<div
  id="container"
>
  <div
    id="textContent"
  >
    <h1
      className="mainHeading"
    >
      Math Magicians
    </h1>
    <h2>
      Let's do some math!
    </h2>
  </div>
  <div
    id="Calculator"
  >
    <div
      id="result"
    >
      <input
        id="result"
        name="result"
        type="text"
        value={null}
      />
    </div>
    <div
      id="leftSide"
    >
      <button
        name="AC"
        onClick={[Function]}
        type="button"
      >
        AC
      </button>
      <button
        name="+/-"
        onClick={[Function]}
        type="button"
      >
        +/-
      </button>
      <button
        name="%"
        onClick={[Function]}
        type="button"
      >
        %
      </button>
      <button
        name="7"
        onClick={[Function]}
        type="button"
      >
        7
      </button>
      <button
        name="8"
        onClick={[Function]}
        type="button"
      >
        8
      </button>
      <button
        name="9"
        onClick={[Function]}
        type="button"
      >
        9
      </button>
      <button
        name="4"
        onClick={[Function]}
        type="button"
      >
        4
      </button>
      <button
        name="5"
        onClick={[Function]}
        type="button"
      >
        5
      </button>
      <button
        name="6"
        onClick={[Function]}
        type="button"
      >
        6
      </button>
      <button
        name="1"
        onClick={[Function]}
        type="button"
      >
        1
      </button>
      <button
        name="2"
        onClick={[Function]}
        type="button"
      >
        2
      </button>
      <button
        name="3"
        onClick={[Function]}
        type="button"
      >
        3
      </button>
      <button
        className="zero"
        name="0"
        onClick={[Function]}
        type="button"
      >
        0
      </button>
      <button
        name="."
        onClick={[Function]}
        type="button"
      >
        .
      </button>
    </div>
    <div
      id="rightSide"
    >
      <button
        className="divide operator"
        name="รท"
        onClick={[Function]}
        type="button"
      >
        รท
      </button>
      <button
        className="multiply operator"
        name="x"
        onClick={[Function]}
        type="button"
      >
        x
      </button>
      <button
        className="minus operator"
        name="-"
        onClick={[Function]}
        type="button"
      >
        -
      </button>
      <button
        className="plus operator"
        name="+"
        onClick={[Function]}
        type="button"
      >
        +
      </button>
      <button
        className="equal operator"
        name="="
        onClick={[Function]}
        type="button"
      >
        =
      </button>
    </div>
  </div>
</div>
`;
